File: lab-4/FiniteStateMachine.cpp
#include "FiniteStateMachine.h"

#include <algorithm>

FiniteStateMachine::FiniteStateMachine(const vector<string> &alphabet, const vector<string> &states,
                                       string initialState,
                                       const vector<Transition> &transitions, const vector<string> &finalStates)
        : alphabet(alphabet),
          states(states),
          initialState(std::move(initialState)),
          transitions(transitions),
          finalStates(finalStates) {

}

const vector<string> &FiniteStateMachine::getAlphabet() const {
    return alphabet;
}

const vector<Transition> &FiniteStateMachine::getTransitions() const {
    return transitions;
}

const vector<string> &FiniteStateMachine::getFinalStates() const {
    return finalStates;
}

const vector<string> &FiniteStateMachine::getStates() const {
    return states;
}

bool FiniteStateMachine::checkSequence(string sequence) {
    string prefix;
    string currentState = initialState;
    bool ok;

    while (!sequence.empty()) {
        ok = false;
        for (const auto& transition: transitions) {
            if (transition.getSourceState() == currentState &&
                transition.getValue() == sequence.substr(0, transition.getValue().length())) {
                prefix += transition.getValue();
                sequence.erase(0, transition.getValue().length());
                currentState = transition.getDestinationState();
                ok = true;
                break;
            }
        }
        if (!ok)
            return false;
    }
    if (find(finalStates.begin(), finalStates.end(), currentState) != finalStates.end())
        return true;
    return false;
}

string FiniteStateMachine::getLongestPrefix(string sequence) {
    string prefix;
    string currentState = initialState;
    bool ok;

    while (!sequence.empty()) {
        ok = false;
        for (const auto& transition: transitions) {
            if (transition.getSourceState() == currentState &&
                transition.getValue() == sequence.substr(0, transition.getValue().length())) {
                prefix += transition.getValue();
                sequence.erase(0, transition.getValue().length());
                currentState = transition.getDestinationState();
                ok = true;
                break;
            }
        }
        if (!ok)
            return prefix;
    }
    return prefix;
}

File: lab-7/finite_state_machine.py
from transition import Transition


class FiniteStateMachine:
    def __init__(self, config_file_name: str):
        self._alphabet = []
        self._states = []
        self._initial_state = None
        self._final_states = []
        self._transitions = []
        self._load_from_file(config_file_name)

    def _load_from_file(self, config_file_name: str) -> None:
        lines = open(config_file_name, "r").readlines()
        self._alphabet = lines[0].rstrip().split(',')
        self._states = lines[1].rstrip().split(',')
        self._initial_state = lines[2].rstrip()
        self._final_states = lines[3].rstrip().split(',')
        self._transitions = []
        for line in lines[4:]:
            words = line.rstrip().split(',')
            self._transitions.append(Transition(words[2], words[0], words[1]))

    def checkSequence(self, sequence: str) -> bool:
        currentState = self._initial_state

        while sequence != '':
            ok = False
            for transition in self._transitions:
                if transition.source_state == currentState and sequence.startswith(
                        transition.value):
                    sequence = sequence.removeprefix(transition.value)
                    currentState = transition.destination_state
                    ok = True
                    break
            if not ok:
                return False
        if currentState in self._final_states:
            return True
        return False
